### ISSUE TITLE
llvm-project/llvm/lib/IR/Constants.cpp:1687: const llvm::APInt &llvm::Constant::getUniqueInteger() const: Assertion `C && isa<ConstantInt>(C) && "Not a vector of numbers!"' failed.

### ISSUE THREAD

**Issue by compnerd** [llvm/llvm-project#57989]
The loop vectorizer hits an assertion when building with optimizations and RVV.  This is an unreduced test case.

```c
// %clang_cc1 -triple riscv64-unknown-linux-musl -emit-obj -target-feature +v -O1 -vectorize-loops %s
// REQUIRES: asserts

void mode_to_string(unsigned mode, char *buf)
{
  for (int i = 0; i < 9; i++) {
    char c = i % 3;
    if (!c && (mode & (1 << ((i / 3) + 9)))) {
      c = "tss"[i / 3];
      if (~mode & (1 << i)) c &= ~0x20;
    }
    buf[9 - i] = c;
  }
}
```

**Comment by topperc** [comment]
Stack trace shows InstCombine.

```
 #7 0x00000000029b776a llvm::Constant::getUniqueInteger() const (./bin/clang+0x29b776a)
 #8 0x0000000002d78eea llvm::InstCombinerImpl::foldSelectIntoOp(llvm::SelectInst&, llvm::Value*, llvm::Value*)::'lambda'(llvm::SelectInst&, llvm::Value*, llvm::Value*, bool)::operator()(llvm::SelectInst&, llvm::Value*, llvm::Value*, bool) const InstCombineSelect.cpp:0:0
 #9 0x0000000002d79208 llvm::InstCombinerImpl::foldSelectIntoOp(llvm::SelectInst&, llvm::Value*, llvm::Value*) (./bin/clang+0x2d79208)
#10 0x0000000002d89349 llvm::InstCombinerImpl::visitSelectInst(llvm::SelectInst&) (./bin/clang+0x2d89349)
```

**Comment by topperc** [comment]
cc @rotateright

**Comment by topperc** [comment]
Proposed fix

```
diff --git a/llvm/lib/IR/Constants.cpp b/llvm/lib/IR/Constants.cpp
index 6eecbb04f2e8..9e84b864e16e 100644
--- a/llvm/lib/IR/Constants.cpp
+++ b/llvm/lib/IR/Constants.cpp
@@ -1682,6 +1682,9 @@ Constant *ConstantVector::getSplatValue(bool AllowUndefs) const {
 const APInt &Constant::getUniqueInteger() const {
   if (const ConstantInt *CI = dyn_cast<ConstantInt>(this))
     return CI->getValue();
+  // Scalable vectors may use a ConstantExpr.
+  if (isa<ConstantExpr>(this))
+    return cast<ConstantInt>(this->getSplatValue())->getValue();
   assert(this->getSplatValue() && "Doesn't contain a unique integer!");
   const Constant *C = this->getAggregateElement(0U);
   assert(C && isa<ConstantInt>(C) && "Not a vector of numbers!");
```

**Comment by topperc** [comment]
> The following assertion probably should be hoisted. However, the proposed fix does seem to fix the assertion failure.

I left the assertion below, because it's the same function I called in the middle of the new return I added for ConstantExpr. The new return is valid for more than just ConstantExpr. I assumed the use of getAggregateElement was a compile time optimization to avoid scanning all elements in release builds so I left that path when possible.

**Comment by compnerd** [comment]
@topperc do you intend to put the change up on phab?

**Comment by topperc** [comment]
> @topperc do you intend to put the change up on phab?

yes I’ll reduce a test this afternoon

**Comment by compnerd** [comment]
```llvm
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n64-S128"
target triple = "riscv64-unknown-linux-musl"

declare <vscale x 8 x i8> @llvm.experimental.vector.reverse.nxv8i8(<vscale x 8 x i8>)
declare <vscale x 8 x i1> @llvm.get.active.lane.mask.nxv8i1.i64(i64, i64)
declare <vscale x 8 x i8> @llvm.masked.gather.nxv8i8.nxv8p0(<vscale x 8 x ptr>, i32 immarg, <vscale x 8 x i1>, <vscale x 8 x i8>)
declare void @llvm.masked.store.nxv8i8.p0(<vscale x 8 x i8>, ptr, i32 immarg, <vscale x 8 x i1>)

define void @mode_to_string(i32 noundef signext %mode) {
entry:
  %broadcast.splatinsert = insertelement <vscale x 8 x i32> poison, i32 %mode, i32 0
  %broadcast.splat = shufflevector <vscale x 8 x i32> %broadcast.splatinsert, <vscale x 8 x i32> poison, <vscale x 8 x i32> zeroinitializer
  %active.lane.mask = call <vscale x 8 x i1> @llvm.get.active.lane.mask.nxv8i1.i64(i64 0, i64 9)
  %wide.masked.gather = call <vscale x 8 x i8> @llvm.masked.gather.nxv8i8.nxv8p0(<vscale x 8 x ptr> zeroinitializer, i32 1, <vscale x 8 x i1> %active.lane.mask, <vscale x 8 x i8> poison)
  %0 = and <vscale x 8 x i32> shufflevector (<vscale x 8 x i32> insertelement (<vscale x 8 x i32> poison, i32 1, i32 0), <vscale x 8 x i32> poison, <vscale x 8 x i32> zeroinitializer), %broadcast.splat
  %1 = icmp eq <vscale x 8 x i32> %0, zeroinitializer
  %2 = and <vscale x 8 x i8> %wide.masked.gather, shufflevector (<vscale x 8 x i8> insertelement (<vscale x 8 x i8> poison, i8 -33, i32 0), <vscale x 8 x i8> poison, <vscale x 8 x i32> zeroinitializer)
  %3 = select <vscale x 8 x i1> %1, <vscale x 8 x i8> %2, <vscale x 8 x i8> %wide.masked.gather
  %predphi34 = select <vscale x 8 x i1> %active.lane.mask, <vscale x 8 x i8> %3, <vscale x 8 x i8> zeroinitializer
  %reverse = call <vscale x 8 x i8> @llvm.experimental.vector.reverse.nxv8i8(<vscale x 8 x i8> %predphi34)
  call void @llvm.masked.store.nxv8i8.p0(<vscale x 8 x i8> %reverse, ptr null, i32 1, <vscale x 8 x i1> zeroinitializer)
  ret void
}
```

Not sure if this can be reduced further.

**Comment by topperc** [comment]
Candidate patch https://reviews.llvm.org/D134670

**Comment by topperc** [comment]
Fixed by 90b695d1f24cd9d6832acf9c7c64ec28a45bb4a1
